core/net/rpl/rpl-icmp6.c

```diff
@@ -996,23 +1444,97 @@ dao_input_nonstoring(void)
     PRINTF("RPL: Sending DAO ACK\n");
     uip_clear_buf();
     dao_ack_output(instance, &dao_sender_addr, sequence,
-        RPL_DAO_ACK_UNCONDITIONAL_ACCEPT);
+                   RPL_DAO_ACK_UNCONDITIONAL_ACCEPT);
   }
 #endif /* RPL_WITH_NON_STORING */
 }
 /*---------------------------------------------------------------------------*/
 static void
 dao_input(void)
 {
+  uint8_t *buffer;
   rpl_instance_t *instance;
   uint8_t instance_id;
+  int sec_len;
+
+#if RPL_SECURITY
+  uint8_t mic_len;      /* n-byte MAC length */
+  rpl_sec_section_t sec_section;
+#endif
 
   /* Destination Advertisement Object */
   PRINTF("RPL: Received a DAO from ");
   PRINT6ADDR(&UIP_IP_BUF->srcipaddr);
   PRINTF("\n");
 
-  instance_id = UIP_ICMP_PAYLOAD[0];
+  buffer = UIP_ICMP_PAYLOAD;
+  sec_len = 0;
+
+#if RPL_SECURITY
+  int pos;
+  uint8_t buffer_length;
+
+  uint8_t timestamp;
+  uint8_t kim;
+  uint8_t lvl;
+  uint32_t counter;
+
+  uint8_t key_index;
+
+  uint8_t ccm_nonce[RPL_NONCE_LENGTH];
+
+  pos = 0;
+
+  rpl_remove_header();
+  buffer = UIP_ICMP_PAYLOAD;
+
+  timestamp = (buffer[pos++] & RPL_TIMESTAMP_MASK) >> RPL_TIMESTAMP_SHIFT;
+  kim = (buffer[pos++] & RPL_KIM_MASK) >> RPL_KIM_SHIFT;
+  lvl = (buffer[pos++] & RPL_LVL_MASK);
+  ++pos;  /* Flags */
+
+/*  Discard packets with time stamp instead of simple counters,
+ *  Only accept packets with pre installed key
+ *  Discard unassigned security levels
+ */
+
+  if((timestamp == 1) || (kim != 0) || (lvl > 3)) {
+    goto discard;
```

Comment:
would be nice to have some counters here, perhaps one for each reason to discard.